### PR TITLE
[PVR][Estuary] Add ParentalRatingCode availability for Skins

### PR DESCRIPTION
**Pull request by DeltaMikeCharlie**
## Description
1) Make the EPG database column 'sParentalRatingCode' available to skins via $INFO[ListItem.parentalratingcode].  This column was added via PR #19395.

2) Display this code in the PVR/TV section of the Estuary skin when it is not null.

## Motivation and context
Add the ability for skins to be able to display the parental rating code (eg: 'PG') instead of the numerical value currently provided by $INFO[ListItem.parentalrating].

## How has this been tested?
Selected rows in the EPG database table were edited to add parental rating code text.  These events were in the past to prevent them from being overwritten.

Modified events displayed the rating code text whilst unmodified events did not.

## What is the effect on users?
If the user connects to a PVR backend that supplies parental rating code information and the addon for communicating with that backend writes said information into the EPG database, the user will be able to see the parental rating code in the footer of the EPG guide window as they navigate the EPG as well as in the detailed event information window.

## Screenshots (if appropriate):

screenshot00001

screenshot00002

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by DeltaMikeCharlie** [comment]
> I've made some formatting changes to meet the current code style. The diffs are available in the following links:
> 
>     * [PR23586-0001-a25b87176d-Add-ParentalRatingCode-availabilit.diff](https://jenkins.kodi.tv/job/BuildMulti-PR/27217/artifact/PR23586-0001-a25b87176d-Add-ParentalRatingCode-availabilit.diff)
> 
>     * [PR23586.diff](https://jenkins.kodi.tv/job/BuildMulti-PR/27217/artifact/PR23586.diff)
> 
> 
> For more information please see our [current code style guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

Jenkins formatting changes applied.

File: xbmc/GUIInfoManager.cpp [hunk 1]
```diff
@@ -6715,6 +6715,14 @@ const infomap container_str[]  = {{ "property",         CONTAINER_PROPERTY },
 ///     @return The parental rating of the list item (PVR).
 ///     <p>
 ///   }
+///   \table_row3{   <b>`ListItem.ParentalRatingCode`</b>,
+///                  \anchor ListItem_ParentalRatingCode
+///                  _string_,
+///     @return The parental rating code (eg: 'PG', etc) of the list item (PVR).
```
Comment (enen92) [inline xbmc/GUIInfoManager.cpp:6721]: Please escale the comma or it will break the table (\,).
Ideally try to build doxygen (folder docs/doxygen) and check the output

File: xbmc/GUIInfoManager.cpp [hunk 2]
```diff
@@ -6862,215 +6870,217 @@ const infomap container_str[]  = {{ "property",         CONTAINER_PROPERTY },
...
-                                  { "samplerate",       LISTITEM_SAMPLERATE },
-                                  { "musicchannels",    LISTITEM_MUSICCHANNELS },
-                                  { "ispremiere",       LISTITEM_IS_PREMIERE },
-                                  { "isfinale",         LISTITEM_IS_FINALE },
-                                  { "islive",           LISTITEM_IS_LIVE },
-                                  { "tvshowdbid",       LISTITEM_TVSHOWDBID },
-                                  { "albumstatus",      LISTITEM_ALBUMSTATUS },
-                                  { "isautoupdateable", LISTITEM_ISAUTOUPDATEABLE },
-                                  { "hdrtype",          LISTITEM_VIDEO_HDR_TYPE },
+const infomap listitem_labels[] = {
```
Comment (enen92) [inline xbmc/GUIInfoManager.cpp:6873]: Please add `// clang-off` at the beginning and `// clang on` at the end to avoid the giant diff

File: addons/skin.estuary/xml/Variables.xml [hunk 3]
```diff
@@ -661,4 +661,7 @@
 		<value condition="VideoPlayer.SubtitlesEnabled">[B]$INFO[VideoPlayer.SubtitlesLanguage][/B]</value>
 		<value>[B]$LOCALIZE[1223][/B]</value>
 	</variable>
+	<variable name="PVRInfoParentalRatingCode">
+		<value condition="!String.IsEmpty(ListItem.parentalratingcode)"> [B][COLOR green]($INFO[ListItem.parentalratingcode])[/COLOR][/B]</value>
```
Comment (enen92) [inline addons/skin.estuary/xml/Variables.xml:665]: Adding colour green to a rating might be misleading. I'd avoid colouring all together and maybe show it decoupled from the genre or it might be misleading as well

Comment (DeltaMikeCharlie) [reply addons/skin.estuary/xml/Variables.xml:665]: @enen92 - Thanks for your input.

The first reason that I coupled the parental rating with the genre is that they are logically related.  A parent may say no to a 'Horror' movie, but yes to the same 'Horror (G)' movie.  Likewise, they may say no to a 'Comedy (R)' movie but maybe accidentally yes to the same 'Comedy' movie.

This would proximity also be beneficial if one day Kodi was able to show parental rating icons from the source system instead of just text.  (Something that I am looking into.)

The second reason was to preserve screen real estate.  Adding a new line and label for the parental rating seemed like a waste of space, especially in the 'MyPVRGuide' view.  There are already several other fields that are strung together onto a single line within this skin.

As for the colour: the (bracket) and colour were to add a small amount of visual separation between the genre and the parental rating.  Green was just a design choice to make it stand out, it could have been anything.

Would you agree to a design compromise?  I make the parental rating code the same dark grey as the labels (not the light grey of the values), but leave it in-line.  Being the same colour as a label would indicate that it is a different field and would suggest that the value displayed is its own label.

Comment (garbear) [reply addons/skin.estuary/xml/Variables.xml:665]: I kinda like the color used as a visual separator, I do the same concept here: https://kodi.wiki/view/Game_add-ons

The colors are of course arbitrary, but the idea is that young kids is green, age 13+ is yellow, and adult-only is red. Color is a way to include a visual separate, in a way that doesn't really matter what the color is, yet is still an indicator to those paying attention.

File: xbmc/GUIInfoManager.cpp (hunk 2 above)
Comment (garbear) [reply xbmc/GUIInfoManager.cpp:6873]: Better would be to have a first commit that applies formatting, and a second commit that gives you a nice beautiful green addition without any surrounding red.

File: addons/skin.estuary/xml/Variables.xml (hunk 3 above)
Comment (DeltaMikeCharlie) [reply addons/skin.estuary/xml/Variables.xml:665]: @garbear - Thanks for your feedback.

I have become aware that some regions use a traffic light system, green/amber/red, to indicate age suitability, so a green '(R)' may be misleading.  It was not my intent to replicate or imply this and I agree that a more neutral colour is desirable.

As part of future development in this area, if icons are added (supplied by the back end), then they can be coloured according to the local laws of that region.

Australia, Brazil, Cyprus, Denmark, Germany, the list goes on, all use coloured icons as part of their classification scheme.

Comment (DeltaMikeCharlie) [reply addons/skin.estuary/xml/Variables.xml:665]: > Would you agree to a design compromise? I make the parental rating code the same dark grey as the labels (not the light grey of the values), but leave it in-line. Being the same colour as a label would indicate that it is a different field and would suggest that the value displayed is its own label.

@enen92 - Do you have any thoughts on this proposal?

Comment (enen92) [reply addons/skin.estuary/xml/Variables.xml:665]: I am fine if the color is removed, not a big deal personally to show it next to the genre. 
However, some review from our PVR guys would be nice (@ksooo @phunkyfish)

Comment (phunkyfish) [reply addons/skin.estuary/xml/Variables.xml:665]: In the screenshots the title colour would work to differentiate it without picking colours.

Comment (DeltaMikeCharlie) [reply addons/skin.estuary/xml/Variables.xml]: The requested changes have been applied.

screenshot00004

screenshot00005

**Review by ksooo — CHANGES_REQUESTED** [review]
Implementation of the GUI info label seems correct. 

What I do not like is the way the skin is extended. Having the parental rating code in braces in the same line as the genre is imo not intuitive. Thus, -1 for this part of the PR.

For the info dialog, why don't you introduce a whole new line "Parent Rating: value"?
For the guide, no idea how to do it differently.

**Comment by DeltaMikeCharlie** [comment]
> Implementation of the GUI info label seems correct.
> 
> What I do not like is the way the skin is extended. Having the parental rating code in braces in the same line as the genre is imo not intuitive. Thus, -1 for this part of the PR.
> 
> For the info dialog, why don't you introduce a whole new line "Parent Rating: value"? For the guide, no idea how to do it differently.

@ksooo - Thanks for your feedback.

I would like to be able to understand your concerns better.  Are you able to more clearly define how you find the proposed solution unintuitive?

Naturally, I disagree that adding the parental rating in parentheses after the genre is not intuitive for the reasons that I have previously stated (6 Aug).  Your comments also seem to be at odds with those previously expressed by your colleagues @enen92 and @phunkyfish.

I would also like to add that using a 'Parenthetical Expression' to convey additional information is a common grammatical practise, at least in English.  Any parent aware of their local classification codes will recognise these as such and anyone else (not interested in ratings) will interpret the text in parentheses as additional information that can be ignored.

Actually, if you read through this entire PR, you will see parenthetical expressions everywhere.  They are used by users (other than me) in their comments and even form part of the PR template 'Screenshots (if appropriate)' and in the 'Types of change' list.  You also added what effectively amounts to 2 parenthetical expressions to the beginning of the title of this PR '[PVR]' and '[Estuary]' albeit in square brackets.

I also found several examples of parenthetical expressions already in use in the Kodi UI.

screenshot00028-small

screenshot00030-small

I'm not trying to be argumentative.  Please help me better understand your concerns.  Even if the rating is moved to a separate line in the Info Dialogue window, we still need to agree on a solution for the Guide Screen given its highly compact nature.  Do you agree to leave the rating at the end of the genre in the Guide Screen?

**Comment by ksooo** [comment]
> I would also like to add that using a 'Parenthetical Expression' to convey additional information is a common grammatical practise, at least in English.

I do agree. As long as the 'Parenthetical Expression' is related to the information before that expression. All the examples for text in parenthesis you gave in the screenshot enriching the information by adding details, like the country for the language, to make the item unique, like giving an example for the selected time format.

For what you suggest I cannot see how the parental rating code relates to the genre, tbh. Not knowing those codes, I have no idea what the text in parenthesis is trying to tell me, really. Without a label making clear what comes next to that label many users will not figure what this text "enriching" the genre is, I'm afraid.

**Comment by DeltaMikeCharlie** [comment]
OK, in summary:

1) The parental rating will be on a separate line on the dialogue screen below the genre.
2) The parental rating will be removed from the EPG screen.

I will be reusing string #563 ("Rating") for the dialogue screen label as it is already used in several other places for this purpose and has translations in 77 languages.

Colours for the label/value will match the other fields such as genre.

If the parental rating string is null, the extra line will not be displayed.

I will make the required changes and submit the amendments.

**Comment by ksooo** [comment]
> I will be reusing string https://github.com/xbmc/xbmc/pull/563 ("Rating") for the dialogue screen label as it is already used in several other places for this purpose

Well, not sure this is the right thing. "Rating" is really generic - could be many different things. I would prefer a new string clearly indicating the purpose.

**Comment by the-black-eagle** [comment]
FWIW, I don't see a problem here.  I'm pretty sure that any parent in the UK knows exactly the meaning of U, PG, 18 etc and that parents around the world also know whatever their equivalent country codes mean.  If you aren't a parent, then it doesn't matter anyway.

Also, my TV and STB both show the rating in the EPG.  EG `44m - PG - Factual`  which I am sure most people can understand perfectly well.  I don't expect to have to flick backwards and forwards between different screens/dialogs to determine if a program is suitable for my child(ren) to watch.

Still, design is very subjective I guess and I have no wish to offend anyone.

**Comment by jjd-uk** [comment]
**Rated** should be used to fit in with other areas of the skin, so example in Movie Info

image

**Rating** would not be correctly as that is normally used to show IMDB vote rating, or user rating.

**Comment by ksooo** [comment]
> Still, design is very subjective I guess and I have no wish to offend anyone.

Absolutely, and as a matter of fact, we will never make everyone happy. And me being the maintainer of the PVR component including its visual manifestation in Estuary, sometimes has just to take decisions. Otherwise we would end up with "everything being displayed everywhere".

**Comment by jjd-uk** [comment]
I don't currently use Kodi PVR so I just noticed we don't currently use media flags for PVR. This info for the Guide could be added with a Rated flag, for example a very quick hack

image

**Comment by ksooo** [comment]
> for example a very quick hack

 Just a hack, indeed. Can we find something better, please? ;-)

**Comment by DeltaMikeCharlie** [comment]
> Well, not sure this is the right thing. "Rating" is really generic

Agreed.

**Comment by DeltaMikeCharlie** [comment]
> In PVR Info dialog then **Rated** should be used to fit in with other areas of the skin, so example in Movie Info dialog

@jjd-uk - Well spotted.  Thank you very much.

The correct string should be #31017 - "Rated".  This is the string that is used in 'DialogVideoInfo.xml' as the label for 'ListItem.Mpaa' which in the property used in the video module.

https://github.com/xbmc/xbmc/blob/1f4ee051090a27f0f37348e8229bdce271ba7036/addons/skin.estuary/xml/DialogVideoInfo.xml#L220

screenshot00031-small

**Comment by DeltaMikeCharlie** [comment]
After a comment left by @the-black-eagle, I am starting to wonder if we may be encountering a simple misunderstanding arising from cultural differences when it comes to the parental rating codes shown in an EPG in various areas of the world.

I spent a few leisurely hours looking at some online EPG examples from around the world.  I looked at examples from: Australia, New Zealand, USA, Canada, Mexico, South Africa, Norway, Germany, France, Italy and UK.

I looked at several dozen examples and only 2 had the label 'Rated' (or similar) displayed.  Most that showed the rating just showed the rating text, although 1 or 2 showed a logo instead of text.  A fair number, maybe 40%, did not show any rating at all.  Several examples had the parental rating in the actual EPG grid text.

There were even examples where multiple fields were strung together on a single line with a field separator character, eg: "1 hr * 2003 * TV-PG * Action/Adventure / Sci-Fi / Space / Adaptation".

Most examples that I found had no field labels of any kind.  They just seemed to assume that people know what type of information to expect in an EPG entry and how to parse that out.  Sometimes, the start date/time was followed by a duration, sometimes it was an end date/time.  Sometimes date and time, sometimes just the time.  Some used bold or colour to break up fields others use a field seperator.  One example used a 'Duration' label, but on the same line after the duration value was a '/' and then the rating logo without a label.

I am **not** suggesting another change.  I think that we should proceed as determined by @ksooo.  However, we need to be aware that in future this design decision may need to be revisited, especially the decision to omit the rating completely from the guide screen.

Ultimately, people can build/choose their own skins if they desire.  The main goal here is to make the rating accessible to the GUI.

**Comment by jjd-uk** [comment]
My view is that Estuary should ideally use logo as that's what everyone is familiar with. For this lots of skins already make use of https://github.com/xbmc/repo-resources/tree/leia/resource.images.classificationicons.colour

Please don't think I am not suggesting that be done for this PR. Just pointing out the future direction I believe we should take if anyone was willing to work on it.

**Comment by ksooo** [comment]
If we want to bring in the PR soon, let's stop debating about the Guide screen now, please. I'm a big fan of the iterative approach. Let's bring in what we all agree on now and not discuss every possible (future) enhancements to death.

File: addons/skin.estuary/xml/Variables.xml [hunk 4]
```diff
@@ -661,4 +661,7 @@
 		<value condition="VideoPlayer.SubtitlesEnabled">[B]$INFO[VideoPlayer.SubtitlesLanguage][/B]</value>
 		<value>[B]$LOCALIZE[1223][/B]</value>
 	</variable>
+	<variable name="PVRInfoParentalRatingCode">
+		<value condition="!String.IsEmpty(ListItem.parentalratingcode)"> [B][COLOR grey]($INFO[ListItem.parentalratingcode])[/COLOR][/B]</value>
```
Comment (ksooo) [inline addons/skin.estuary/xml/Variables.xml:665]: I think the formatting (COLOR, bold) should not be part of the variable. This may hinder using the var from windows where colours/stype/... is different.

File: addons/skin.estuary/xml/MyPVRGuide.xml [hunk 5]
```diff
@@ -208,7 +208,7 @@
 							<width>40%</width>
 							<height>30</height>
 							<align>right</align>
-							<label>$INFO[ListItem.Genre,[COLOR grey]$LOCALIZE[515]:[/COLOR] ]</label>
+							<label>$INFO[ListItem.Genre,[COLOR grey]$LOCALIZE[515]:[/COLOR] ]$VAR[PVRInfoParentalRatingCode]</label>
```
Comment (ksooo) [inline addons/skin.estuary/xml/MyPVRGuide.xml:211]: Please remove.

**Review by ksooo — CHANGES_REQUESTED** [review]

**Comment by ksooo** [comment]
@DeltaMikeCharlie I think we should add the new label to channels window's right hand side info panel:

screenshot00004

**Comment by DeltaMikeCharlie** [comment]
> My view is that Estuary should ideally use the logo as that's what everyone is familiar with. For this lots of skins already make use of https://github.com/xbmc/repo-resources/tree/leia/resource.images.classificationicons.colour
> 
> Please don't think I am suggesting that be done for this PR. Just pointing out the future direction I believe we should take if anyone was willing to work on it.

@jjd-uk - Perhaps you should start a thread on the Kodi forum so that we can continue this discussion there in more detail.

In short:  Here is a post that I made on the tvheadend forum discussing the work that I have done there on converting the age rating values transmitted in the DVB EIT into usable parental rating labels and icons: https://tvheadend.org/boards/5/topics/50356

**Comment by DeltaMikeCharlie** [comment]
> @DeltaMikeCharlie this needs a rebase

@ksooo - I'm not very experienced with GitHub, so I need some advice on how to do this correctly.

At the start, I cloned the xbmc master repository into my private repository on GitHub.  On GitHub, I also created a branch in my repository to add this feature.  I then pulled that feature branch onto my local development environment where I made my changes before pushed them back to GitHub in order to create the PR.

So good so far....

How do I rebase my local feature branch to get it up-to-date with the xbmc master whilst still maintaining my current local changes?

Do I need to first resync my xbmc repository to the xbmc master repository in GitHub and then rebase my local feature branch with my GitHub xbmc repository?  I'm guessing that I will also need to 'stash' my local changes before rebasing in order to preserve them.

File: addons/skin.estuary/xml/Variables.xml (hunk 4 above)
Comment (DeltaMikeCharlie) [reply addons/skin.estuary/xml/Variables.xml:665]: Actually, I deleted the variable completely.  The variable was only needed so that the parental rating could appear to be appended to the genre string.  Now that the parental rating has been removed from the EPG screen and given its own line plus label in the dialogue screen, I can just refer to it directly via $INFO like any other property.

If another window needs access to the parental rating, like the PVRInfoPanel, then it can also use $INFO with the property name.

File: addons/skin.estuary/xml/MyPVRGuide.xml (hunk 5 above)
Comment (DeltaMikeCharlie) [reply addons/skin.estuary/xml/MyPVRGuide.xml:211]: Done.

01-EPG Screen (Removed)

**Comment by DeltaMikeCharlie** [comment]
> @DeltaMikeCharlie I think we **also** should add the new label to channels window's right hand side info panel:
> 
> screenshot00004
> 
> Look for `PVRInfoPanel` in Includes_PVR.xml, please. There the change needs to be done.

Done.

03-Channel List (Added, own line)

**Comment by DeltaMikeCharlie** [comment]
The parental rating has also been moved to its own line with a label in the dialogue window.

02-EPG Dialogue (Own line)

I just need to do the rebase and push the update and it should be good to go.

**Comment by the-black-eagle** [comment]
@DeltaMikeCharlie Does this help any -> https://github.com/xbmc/xbmc/blob/master/docs/GIT-FU.md#6-syncing-branches

As long as your local changes are committed, there is no need to stash anything.  Switch to master, fetch the current upstream branch etc then switch back to your local feature branch and `git rebase -i origin/master` (assuming you have pushed the new master branch to your remote github).  git will mark in the files where the conflicts are.  Edit the files to how they should be and then `git add <edited file>` when all re-added run `git rebase --continue`.  When git tells you it has successfully rebased, force push your branch with `git push -f origin`.  Or at least, that's how I do it.  `git add -p` is helpful just to make sure you really are adding the right stuff.

File: addons/skin.estuary/xml/Variables.xml (hunk 4 above)
Comment (ksooo) [reply addons/skin.estuary/xml/Variables.xml:665]: > Actually, I deleted the variable completely. 

Great. I was about to suggest exactly this.

**Comment by DeltaMikeCharlie** [comment]
> @DeltaMikeCharlie Does this help any -> https://github.com/xbmc/xbmc/blob/master/docs/GIT-FU.md#6-syncing-branches
> 
> As long as your local changes are committed, there is no need to stash anything. Switch to master, fetch the current upstream branch etc then switch back to your local feature branch and `git rebase -i origin/master` (assuming you have pushed the new master branch to your remote github). git will mark in the files where the conflicts are. Edit the files to how they should be and then `git add <edited file>` when all re-added run `git rebase --continue`. When git tells you it has successfully rebased, force push your branch with `git push -f origin`. Or at least, that's how I do it. `git add -p` is helpful just to make sure you really are adding the right stuff.

@the-black-eagle - Thanks for your help.  I read that document link earlier today and it partially helps, but not completely.  I still have a question about 1 thing:

If I switch to my local master branch and then fetch the current upstream branch as you suggest, then I don't see what difference that will make unless I first synchronise my upstream master branch with the xbmc master.

I was thinking that I would have to:

1. Resync my GitHub master to the xbmc GitHub master.
2. Switch to my local master.
3. Fetch from my GitHub master.
4. Switch back to my local feature branch.
5. Rebase and resolve conflicts.
6. Force push of my local feature branch back to my GitHub feature branch.

Have I missed/misunderstood something?

**Comment by the-black-eagle** [comment]
I have a bash alias set up for updating my mater branch.  It switches to my local master, fetches upstream master, merges the changes then pushes the result to my remote.  
`alias updatemaster='git checkout master && git fetch upstream && git merge upstream/master && git push origin'`

You might have to subsitute `upstream` if you have called it something else, but basically, `upstream` points at the team repo here, and `origin` points at my repo.

Basically, that checks out your local master, fetches the current upstream changes, merges them and pushes the result.  Your local and remote master will then match the teams.  Then you rebase on top of that.

**Comment by phunkyfish** [comment]
@DeltaMikeCharlie here is the guide I use to setup my dev repos. It should help. Once setup correctly the commands @the-black-eagle gave you should work. Obviously you have step 1, done already.

1 - After creating a fork on github make a local clone:

git clone https://github.com/YOUR-USERNAME/repo-name

2 - Add upstream as a remote

Type git remote -v and press Enter. You'll see the current configured remote repository for your fork.
$ git remote -v
> origin  https://github.com/YOUR_USERNAME/YOUR_FORK.git (fetch)
> origin  https://github.com/YOUR_USERNAME/YOUR_FORK.git (push)
Type git remote add upstream, and then paste the URL you copied in Step 2 and press Enter. It will look like this:

	$ git remote add upstream [https://github.com/](https://github.com/octocat/Spoon-Knife.git)ORIGINAL_OWNER/ORIGINAL_REPOSITORY.git

To verify the new upstream repository you've specified for your fork, type git remote -v again. You should see the URL for your fork as origin, and the URL for the original repository as upstream.
$ git remote -v
> origin    https://github.com/YOUR_USERNAME/YOUR_FORK.git (fetch)
> origin    https://github.com/YOUR_USERNAME/YOUR_FORK.git (push)
> upstream  https://github.com/ORIGINAL_OWNER/ORIGINAL_REPOSITORY.git (fetch)
> upstream  https://github.com/ORIGINAL_OWNER/ORIGINAL_REPOSITORY.git (push)
3 - Now fetch the upstream so you can track upstream master:

	git fetch upstream

4 - Tracking master branch from upstream:

	git branch -u upstream/master

Or, if local branch foo is not the current branch:

	git branch -u upstream/master master

Or, if you like to type longer commands, these are equivalent to the above two:

 git branch --set-upstream-to=upstream/master

 git branch --set-upstream-to=upstream/master master

5 - No pushing to upstream!

	git remote set-url upstream --push “No way buddy, you can’t push there"

**Comment by DeltaMikeCharlie** [comment]
@phunkyfish, @the-black-eagle - Thank you both.

The 'git remote -v' command helped me see that my misunderstand was actually with nomenclature.  My GitHub clone of xbmc is the 'origin' whereas the master xbmc repository is the 'upstream'.

I'll try the rebase in the morning when my mind is fresher.

**Comment by DeltaMikeCharlie** [comment]
@the-black-eagle
@phunkyfish 

Thanks for your help, the rebase went relatively smoothing.

However, when I rebuilt my local branch after the rebase, it built cleanly but Kodi produced a segmentation fault when executed.  I saw this on another project recently (branched after this one), but this one seemed OK until the rebase.

I tried to create a clean build directory to rebuild my local branch from scratch and the build commands gave errors.

When I was resolving the merge conflicts, there were so few and they were so trivial that I would have been better off just applying the changes manually and then pushing that back to GitHub.

Alas, now that my build system appears to be one giant potato, I'm going to build a new Kodi development VM from scratch and start again.  This may take a day or 2.  I have been wanting to rebuild this for a while, it's a shame that my hand was forced this close to finishing off this PR.

Is is possible to undo the rebase?

**Comment by ksooo** [comment]
Looking at the number of changed lines now it seems something is not as it should be after the rebase. Far too many changed lines in CGUIInfoManager…

**Comment by ksooo** [comment]
May I suggest that I take over and do the chases for you. Could be that this saves some time for all of as. Just asking.

**Comment by ksooo** [comment]
If you have local backup files, you could just do a “git reset —hard upstream/master” on your branch. Then re-apply your changes, commit and “git push origin -ff your-branch-name

**Comment by DeltaMikeCharlie** [comment]
> May I suggest that I take over and do the rebase for you.

@ksooo - Thank you.  I appreciate the offer, but I really need to learn how to do a rebase for myself.  Do you have a deadline for this PR to be merged?

> If you have local backup files, you could just do a “git reset —hard upstream/master” on your branch. Then re-apply your changes, commit and “git push origin -ff your-branch-name”

I'm going to try this and see if it works.

I have built a new Ubuntu VM and got Kodi to compile and run once already.  If the hard reset above fails, it may be quicker to keep going with getting the new VM working.  I noticed that even though I ran kodi-x11, it seems to be referring to the monitors as XWAYLANDn.  The other SEGFAULT that I was having was on startup as the logo was displayed.  Perhaps my old development environment was not up-to-date enough to handle the latest Kodi.

**Comment by DeltaMikeCharlie** [comment]
> Looking at the number of changed lines now it seems something is not as it should be after the rebase. Far too many changed lines in CGUIInfoManager…

I think that there are 2 things happening here:  1) There were some Jenkins-mandated changes applied earlier on on my PR and 2) I inserted something into the middle of listitem_labels[] and someone else inserted something at the end so the whole {} block was flagged as conflicting.  However, there is really no conflict and the new lines can simply be inserted where they are.

**Comment by ksooo** [comment]
> Do you have a deadline for this PR to be merged?

No, but I feel like you keep some people here quite busy to teach you git, to read and think through what write here. Me just doing it would save us all a bit of our time.

**Comment by DeltaMikeCharlie** [comment]
Two steps forward and one step back.

I finally got my new development environment to build Kodi into an executable that did not SEGFAULT immediately and then push my changes to GitHub.

There was still a single line merge conflict that I resolved manually on GitHub.

Unfortunately, the automatic build process failed stating 'ERROR: This PR contains 1 merge commit(s)'.

Guidance on resolving this error would be appreciated.

**Comment by howie-f** [comment]
> ERROR: This PR contains 1 merge commit(s)

rebasing your development branch on upstream/master will make the merge commit disappear.

File: xbmc/GUIInfoManager.cpp [hunk 6]
```diff
@@ -6871,217 +6871,217 @@ const infomap container_str[]  = {{ "property",         CONTAINER_PROPERTY },
...
-    {"islive", LISTITEM_IS_LIVE},
-    {"tvshowdbid", LISTITEM_TVSHOWDBID},
-    {"albumstatus", LISTITEM_ALBUMSTATUS},
-    {"isautoupdateable", LISTITEM_ISAUTOUPDATEABLE},
-    {"hdrtype", LISTITEM_VIDEO_HDR_TYPE},
+const infomap listitem_labels[]= {{ "thumb",            LISTITEM_THUMB },
+                                  { "icon",             LISTITEM_ICON },
+                                  { "actualicon",       LISTITEM_ACTUAL_ICON },
+                                  { "overlay",          LISTITEM_OVERLAY },
+                                  { "label",            LISTITEM_LABEL },
```
Comment (garbear) [inline xbmc/GUIInfoManager.cpp]: This blows git history for a bunch of lines, better to add `//clang-format off` before and `//clang-format on` after?

**Review by ksooo — APPROVED** [review]
Finally. ;-) Thank you so much for your patience.

**Comment by DeltaMikeCharlie** [comment]
> Finally. ;-) Thank you so much for your patience.

No @ksooo, thank you for your help, guidance and tolerance.

Also thanks to @garbear, @enen92, @the-black-eagle, @phunkyfish and @howie-f as well for their suggestions along the way.

I am only speculating, but I think that my original development system may have been faulty in some way.  Perhaps after the rebase, some of the dependencies became comparatively out of date in a way that allowed Kodi to compile and link cleanly, but immediately sigfault when run.  Building a new development VM and dependencies from scratch seems to have resolved the issue.

**Comment by phunkyfish** [comment]
I think it 6 years ago now that @ksooo coached me on `git` and `c++` when I first started contributing!

You have to start somewhere 😉
